docs/resources/server/tags.md

```diff
@@ -1,83 +1,240 @@
+```java
+@SubscribeEvent
+public static void gatherData(GatherDataEvent event) {
+    PackOutput output = generator.getPackOutput();
+    CompletableFuture<HolderLookup.Provider> lookupProvider = event.getLookupProvider();
+    ExistingFileHelper existingFileHelper = event.getExistingFileHelper();
+
+    // other providers here
+    event.getGenerator().addProvider(
+        event.includeServer(),
+        new MyBlockTagsProvider(output, lookupProvider, existingFileHelper)
+    );
+}
+```
 
-ResourceKey<VillagerType> villagerTypeKey = /*...*/;
-boolean isInVillagerTypeGroup = BuiltInRegistries.VILLAGER_TYPE.getHolder(villagerTypeKey).map(holder -> holder.is(myVillagerTypeTag)).orElse(false);
+`ItemTagsProvider` has an additional helper method called `#copy`. It is intended for the common use case of item tags mirroring block tags:
+
+```java
+// In an ItemTagsProvider's #addTags method, assuming types TagKey<Block> and TagKey<Item> for the two parameters.
+copy(EXAMPLE_BLOCK_TAG, EXAMPLE_ITEM_TAG);
 ```
 
-## Conventions
+### Custom Tag Providers
 
-There are several conventions that will help facilitate compatibility in the ecosystem:
+To create a custom tag provider for a custom [registry], or for a vanilla or NeoForge registry that doesn't have a tag provider by default, you can also create custom tag providers like so (using recipe type tags as an example):
 
-- If there is a Vanilla tag that fits your registry object, add it to that tag. See the [list of Vanilla tags][taglist].
-- If there is a group of something you feel should be shared by the community, use the `c` namespace instead of your mod id.
-    - The `c` namespace is a common namespace for NeoForge and Fabric.
-    - Tags should be sorted into subdirectories according to their type (e.g. `c:ingots/iron`, `c:nuggets/brass`, etc.).
-- If there is a NeoForge tag that fits your registry object, add it to that tag. The list of tags declared by NeoForge can be seen on [GitHub][neoforgetags].
-- Tag naming conventions should follow Vanilla conventions.(e.g. plural instead of singular: `minecraft:logs`, `minecraft:saplings`).
+```java
+public class MyRecipeTypeTagsProvider extends TagsProvider<RecipeType<?>> {
+    // Get parameters from GatherDataEvent.
+    public MyRecipeTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, ExistingFileHelper existingFileHelper) {
+        // Second parameter is the registry key we are generating the tags for.
+        super(output, Registries.RECIPE_TYPE, lookupProvider, ExampleMod.MOD_ID, existingFileHelper);
+    }
+    
+    @Override
+    protected void addTags(HolderLookup.Provider lookupProvider) { /*...*/ }
+}
+```
 
-## Using Tags in Recipes and Advancements
+If desirable and applicable, you can also extend `IntrinsicHolderTagsProvider<T>` instead of `TagsProvider<T>`. This additionally requires a function parameter that returns a resource key for a given object. Using attribute tags as an example:
```

Comment:
Explain why you would use `IntrinsicHolderTagsProvider`! It's because then you can pass in the object to tags rather than the key!